lib/multiplexer/multiplexer.go

```diff
@@ -554,3 +596,34 @@ func detectProto(r *bufio.Reader) (Protocol, error) {
 
 	return ProtoUnknown, trace.BadParameter("multiplexer failed to detect connection protocol, first few bytes were: %#v", in)
 }
+
+type PROXYHeaderSigner interface {
```

Comment:
You mean extract everything relate to proxylines, including implementation of reading/writing them, into some sub package `proxyline` in multiplexer folder?

Reply:
I see, so basically something separate PROXY recognition from multiplexer and provide it with a listener that already internally did detection on the connection to check for PROXY protocol? Basically creating `PROXYEnabledListener` before creating `mux` instance. Not sure how good it would be in practice, would it make thing more less complicated or more. But I played with separating proxy related stuff into separate sub package and it looks like it can be done, without cycling references (probably not without a little bit of code duplication in tests). It would be first step into logical separation, but PR would be pretty large by itself, so I wouldn't want to mix it into current PR 🤔

Reply:
I've opened an issue for decoupling PROXY from multiplexer: https://github.com/gravitational/teleport/issues/21835

Reply:
I tired it and it just pulls one thing after another (like ProxyLine struct, and all it's methods) including tests and it just becomes 1000+ line changes thing on itself. I think it's better to have clean PR dedicated to that split.